Implement catch-up phase

AndrejMitrovic opened the pull request:
It's not ready for merging yet, but it can be looked at.

This still needs a bit of refactoring, separating atomic changes into their own commits.

AndrejMitrovic committed:
Make addNewBlock() public

It will be used to add blocks received
from other nodes.

AndrejMitrovic commented:
Also this only implements catch-up during booting up. The other definitions of done in the Catchup Phase Issue need to be solved.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -76,3 +77,47 @@ unittest
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.digest.sha;
```
Should be removed

AndrejMitrovic replied:
damn where did that sneak in from lol

AndrejMitrovic committed:
Remove unused import

AndrejMitrovic committed:
Implement getBlockHeight() wrapper in RemoteNode

AndrejMitrovic committed:
Implement getBlocksFrom() wrapper in RemoteNode

AndrejMitrovic committed:
Implement catchup phase of a node

After a node connects to all its peers,
it will retrieve the latest blocks and
apply them to its ledger.

The node will keep periodically trying
to retrieve the latest blocks from its
connected peers.

AndrejMitrovic commented:
I've atomic'ed (atomized?) the commits, it should look much nicer now.

I probably need to think of edge cases to add to the unittests.

Geod24 commented on source/agora/node/RemoteNode.d:
```diff
@@ -170,4 +170,25 @@ class RemoteNode
+        catch (Exception ex)
+        {
+            logError(ex.message);
+            return 0;
```
`ulong.max`

AndrejMitrovic replied:
Good point, block height is zero for genesis blocks. I'll add a test-case. Also like I said before I dislike this class, but I plan on making it better at a later point. There's no retry mechanism for example.

Geod24 replied:
Ping

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -183,10 +188,14 @@ public final class TestNode (Net) : Node!Net, TestAPI
     public this (Config config)
     {
         this.mem_metadata = new MemMetadata;
+        this.ledger = new Ledger;
```
That code doesn't make sense to me. You instantiate a new `Ledger` just to override it with itself in the parent constructor. You also shadow the parent `ledger` variable ?

AndrejMitrovic replied:
The shadowing was an oversight. I'll simplify this a bit tomorrow.

AndrejMitrovic commented:
Update: Added periodic catch-up phase, and added a network test for it. This completes all of the tasks and should cover everything in the definition of done.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -175,6 +177,9 @@ public interface TestAPI : API
     public abstract void metaAddPeer(string peer);
+
+    ///
+    public abstract void addBlockToLedger (Block block);
```
Why do you need this ? I don't see an override anywhere ?

AndrejMitrovic replied:
Well, the question is how do you get access to the Ledger from the unittest. As it's required right now. edit: reworded

Geod24 replied:
You never access the internals of a node at the moment.

AndrejMitrovic replied:
Well the first test is:

Instantiate the nodes. Don't start networking yet. One node (node A) needs to have 100 blocks, the 3 are empty.

Then, make them connect to each other. The 3 other nodes will retrieve blocks from the 1st node on boot.

But to make this work, I need to add those blocks to node A before it connects to the 3 other nodes.

AndrejMitrovic replied:
To be more specific here. I can't just use the `sendTransaction` API here because this will trigger the gossip protocol to send the transaction over the whole network, and then each node will create their own block. So by the time the node tries to retrieve the blocks from the other nodes, it will already have all the blocks.

Delaying the node might not help. I would need some way to temporarily disable the gossip protocol, but then this becomes very messy to have all these special cases for unittesting.

The more features the node has, the more complicated it is to test a specific feature in the node without triggering an avalanche of other features.

AndrejMitrovic replied:
I will be able to replace this code once LocalRest has the ability to filter requests. Then I can do something like:

- Use the sendTransaction API to one node
- Make the other nodes ignore transaction gossiping
- The other nodes will then catch-up using the periodic `getBlocksFrom` call.

This is way nicer than messing with Node's internals.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -118,3 +118,48 @@ unittest
+{
+    import core.thread;
+    import std.conv;
+    import vibe.core.log;
```
Two leftover imports

Geod24 reviewed (changes requested):
I would merge the last two commits. Also, do you want the ability to arbitrarily delay a node ? Because I can give you that power easily in `LocalRest`.

AndrejMitrovic commented:
I've squashed the two commits.

I'm thinking about how to get rid of those methods in the TestAPI now.

AndrejMitrovic commented:
> Also, do you want the ability to arbitrarily delay a node ? Because I can give you that power easily in LocalRest.

Hmm.. that might be very useful.

AndrejMitrovic commented:
All the Mac builds fail. Hmm..

AndrejMitrovic commented:
It's a timing issue.

AndrejMitrovic commented:
Ok I made a more stable version of the tests that shouldn't suffer too much from timing issues. I added polling and a timeout of 4 seconds.

AndrejMitrovic commented:
Very peculiar that macOS dmd 2.085.1 seems to consistently be failing now..

AndrejMitrovic commented:
~@Geod24 I'm having consistent time-outs in the MacOS build: https://travis-ci.com/bpfkorea/agora/jobs/214998228

It seems it gets stuck trying to install DMD & Dub.~

And it's green.

Geod24 commented:
I guess you want to rebase ?

AndrejMitrovic commented:
> I guess you want to rebase ?

Ah I thought this one was rebased.

AndrejMitrovic commented:
I mean merged.

AndrejMitrovic commented:
Rebased.

Geod24 commented on source/agora/node/Network.d:
```diff
@@ -137,6 +140,72 @@ public class NetworkManager
+            // periodic task
+            while (1)
+            {
+                logInfo("Retrieving latest blocks..");
```
I think this logging should only happen if we are actually catching up

Geod24 commented on source/agora/node/RemoteNode.d:
```diff
@@ -170,4 +171,55 @@ class RemoteNode
+
+    ***************************************************************************/
+
+    public ulong getBlockHeight ( )
```
```suggestion
    public ulong getBlockHeight ()
```

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -22,13 +23,24 @@ import agora.common.Hash;
 import agora.test.Base;
 
+/// Returns: true if all node's block heights are equal to the provided height
+private bool allHeightsEqual ( TestAPI[] nodes, size_t height )
```
```suggestion
private bool allHeightsEqual (TestAPI[] nodes, size_t height)
```

Geod24 reviewed (changes requested):
As discussed IRL: Bad rebase, some code is missing

AndrejMitrovic commented:
Fixed the suggested changes.

AndrejMitrovic commented:
Rebased to use new LocalRest sleep() feature.

However it's a bit ugly because now I have to store `RemoteAPI!TestAPI` in `TestNetworkManager`, not just `TestAPI`. And then there's the issue of `RemoteAPI!TestAPI[]` not implicitly converting to `TestAPI[]`.

AndrejMitrovic commented:
Edit: Pushed a simplification https://github.com/bpfkorea/agora/compare/3c38af6c675d18cab84aa8c89c702d6b3cf8bb7c..bf03c16e9bc1267934450648f55a9f583ffbc9e1

Geod24 commented on source/agora/node/Network.d:
```diff
@@ -137,6 +140,72 @@ public class NetworkManager
+    ***************************************************************************/
+
+    private void getBlocksFrom (ulong block_height,
+        void delegate(Block[]) @safe onReceivedBlocks)
```
```suggestion
        scope void delegate(Block[]) @safe onReceivedBlocks)
```

Geod24 commented on source/agora/node/Node.d:
```diff
@@ -174,4 +176,18 @@ public class Node (Network) : API
+
+    protected Ledger getLedger ()
+    {
+        return new Ledger();
```
I don't think you need this anymore ?

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -71,3 +83,79 @@ unittest
+{
+    import core.thread;
+    import std.algorithm;
+    import std.format;
```
Those 2 imports are not necessary anymore

AndrejMitrovic replied:
I need `each` still.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -22,13 +23,24 @@ import agora.common.Hash;
 import agora.test.Base;
 
+/// Returns: true if all node's block heights are equal to the provided height
+private bool allHeightsEqual (API)(API[] nodes, size_t height)
```
Suggestion: `isConsistent` and check the block headers for equality ?

AndrejMitrovic replied:
Ok.

AndrejMitrovic commented:
Updated with suggested changes.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -71,3 +83,78 @@ unittest
+    import core.thread;
+    import std.algorithm;
+    import std.conv;
+    import std.format;
```
conv and format superfluous

Geod24 reviewed (approved):
LGTM

AndrejMitrovic commented:
Pushed a small update.
